### PR TITLE
PTX-2757 CSI validation improvements and regression test

### PR DESCRIPTION
**Pull request by nikolaypopov**
CSI basic regression test and various improvements for validation of CSI components and images

* Added CSI basic regression test that cover multiple CSI test cases
             CSI - Validate CSI is enabled by default
             CSI - Enable CSI on already deployed non CSI PX cluster 
             CSI - Disable CSI on PX cluster 
             CSI - Validate px-csi-ext deployment and pods 
             CSI - Validate csi-node-driver-registrar container inside PX pod
             CSI - Delete px-csi-ext pods and validate they get redeployed 
             CSI - Delete portworx pods and validate they get redeployed with the CSI container healthy 
* Added more in-depth validation for CSI components and images

Other fixes and improvements that were blockers and had to be included in this PR:
* Added fix for ConstructVersionURL() to take k8s version, otherwise we are getting wrong images for various components
* Slightly changed Upgrade test

### CODE REVIEW

**Review by piyush-nimbalkar — COMMENTED** [review]
This looks nice! Good coverage on the test cases too.
Some minor comments and suggestions.

File: pkg/util/test/util.go [hunk 1]
```diff
@@ -867,32 +876,251 @@ func validateComponents(pxImageList map[string]string, cluster *corev1.StorageCl
...
 }
 
-func validatePodsByName(namespace, name string, timeout, interval time.Duration) error {
-	listOptions := map[string]string{"name": name}
+func validatePodsByLabel(namespace, label, name string, timeout, interval time.Duration) error {
+	listOptions := map[string]string{label: name}
 	return ValidatePods(namespace, listOptions, timeout, interval)
 }
 
+// DeletePodsByDeployment deletes pods for the given deployment name and namespace
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Moved to sched-ops and update sched-ops

File: pkg/util/test/util.go [hunk 2]
```diff
@@ -867,32 +876,251 @@ func validateComponents(pxImageList map[string]string, cluster *corev1.StorageCl
...
+				}
+			}
+		}
+	}
 	return nil
 }
 
-func validatePodsByName(namespace, name string, timeout, interval time.Duration) error {
-	listOptions := map[string]string{"name": name}
+func validatePodsByLabel(namespace, label, name string, timeout, interval time.Duration) error {
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Removed

File: pkg/util/test/util.go [hunk 3]
```diff
@@ -867,32 +876,251 @@ func validateComponents(pxImageList map[string]string, cluster *corev1.StorageCl
...
+func validateCsiExtImages(namespace string, pxImageList map[string]string) error {
+	var csiProvisionerImage string
+	var csiSnapshotterImage string
+	var csiResizerImage string
+
+	logrus.Debug("Validating CSI container images inside px-csi-ext pods")
+
+	// Get px-csi-ext pods
+	listOptions := map[string]string{"app": "px-csi-driver"}
+	pods, err := coreops.Instance().GetPods(namespace, listOptions)
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go:981]: Done

File: pkg/util/test/util.go [hunk 4]
```diff
@@ -867,32 +876,251 @@ func validateComponents(pxImageList map[string]string, cluster *corev1.StorageCl
...
+		return fmt.Errorf("Failed to find image for csiNodeDriverRegistrar")
+	}
+
+	// Go through each pod and find all container and match images for each container
+	for _, pod := range pods.Items {
+		for _, container := range pod.Spec.Containers {
+			if container.Name == "csi-node-driver-registrar" {
+				if container.Image != csiNodeDriverRegistrar {
+					return fmt.Errorf("Found container %s, expected image: %s, actual image: %s", container.Name, csiNodeDriverRegistrar, container.Image)
+				}
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go:963]: Done

File: pkg/util/test/util.go [hunk 5]
```diff
@@ -867,32 +876,251 @@ func validateComponents(pxImageList map[string]string, cluster *corev1.StorageCl
...
+	} else if !csi {
+		logrus.Debug("CSI is disabled in StorageCluster")
+
+		// Validate px-csi-ext deployment doesn't exist
+		retCsiDpObject, err := appops.Instance().GetDeployment(pxCsiDp.Name, pxCsiDp.Namespace)
+		if err != nil || retCsiDpObject == nil {
+			logrus.Debugf("CSI Deployment %s is nil as expected, Err: %v", pxCsiDp.Name, err)
+		} else {
+			return fmt.Errorf("failed to validate CSI. Found deployment %s when it was not suppose to be deployed", pxCsiDp.Name)
+		}
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go:927]: Done

File: pkg/util/test/util.go [hunk 6]
```diff
@@ -867,32 +876,251 @@ func validateComponents(pxImageList map[string]string, cluster *corev1.StorageCl
...
+		// Validate px-csi-ext deployment and pods
+		if err := validateDeployment(pxCsiDp, timeout, interval); err != nil {
 			return err
 		}
+
+		// Validate CSI container images inside px-csi-ext pods
+		if err := validateCsiExtImages(cluster.Namespace, pxImageList); err != nil {
+			return err
+		}
+	} else if !csi {
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Done

File: test/integration_test/basic_test.go [hunk 7]
```diff
@@ -273,3 +300,98 @@ func testInstallWithTelemetry(t *testing.T, cluster *op_corev1.StorageCluster) {
...
+		logrus.Infof("Validate StorageCluster %s", cluster.Name)
+		err = testutil.ValidateStorageCluster(pxSpecImages, cluster, defaultValidateDeployTimeout, defaultValidateDeployRetryInterval, true, "")
+		require.NoError(t, err)
+
+		// Delete portworx oci-mon pods and validate they get redeploymed
+		logrus.Info("Delete portworx pods and validate they get re-deployed")
+		err = testutil.DeletePodsByLabel(cluster.Namespace, "name", "portworx")
+		require.NoError(t, err)
+		err = testutil.ValidateStorageCluster(pxSpecImages, cluster, defaultValidateDeployTimeout, defaultValidateDeployRetryInterval, true, "")
+		require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply test/integration_test/basic_test.go]: Done

File: test/integration_test/basic_test.go [hunk 8]
```diff
@@ -273,3 +300,98 @@ func testInstallWithTelemetry(t *testing.T, cluster *op_corev1.StorageCluster) {
...
+
+		// Delete px-csi-ext pods and validate they get redeploymed
+		logrus.Info("Delete px-csi-ext pods and validate they get re-deployed")
+		err = testutil.DeletePodsByDeployment("px-csi-ext", cluster.Namespace)
+		require.NoError(t, err)
+		err = testutil.ValidateStorageCluster(pxSpecImages, cluster, defaultValidateDeployTimeout, defaultValidateDeployRetryInterval, true, "")
+		require.NoError(t, err)
+
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply test/integration_test/basic_test.go]: Doing it only once

File: test/integration_test/basic_test.go [hunk 9]
```diff
@@ -273,3 +300,98 @@ func testInstallWithTelemetry(t *testing.T, cluster *op_corev1.StorageCluster) {
...
+		// Deploy cluster
+		cluster, err = createStorageCluster(cluster)
+		require.NoError(t, err)
+
+		// Validate cluster deployment
+		logrus.Infof("Validate StorageCluster %s", cluster.Name)
+		err = testutil.ValidateStorageCluster(pxSpecImages, cluster, defaultValidateDeployTimeout, defaultValidateDeployRetryInterval, true, "")
+		require.NoError(t, err)
+
+		// Delete px-csi-ext pods and validate they get redeploymed
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply test/integration_test/basic_test.go]: Fixed comment

File: test/integration_test/basic_test.go [hunk 10]
```diff
@@ -97,6 +98,20 @@ var testStorageClusterBasicCases = []TestCase{
 		ShouldSkip: func() bool { return false },
 		TestFunc:   InstallWithTelemetry,
 	},
+	{
+		TestName:        "InstallWithCSI",
+		TestrailCaseIDs: []string{},
+		TestSpec: func(t *testing.T) interface{} {
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply test/integration_test/basic_test.go]: Done

File: test/integration_test/basic_test.go [hunk 11]
```diff
@@ -243,3 +270,79 @@ func testInstallWithTelemetry(t *testing.T, cluster *corev1.StorageCluster) {
...
+
+		var err error
+		testSpec := tc.TestSpec(t)
+		cluster, ok := testSpec.(*corev1.StorageCluster)
+		require.True(t, ok)
+
+		// Create and validate StorageCluster
+		cluster = ci_utils.DeployAndValidateStorageCluster(cluster, ci_utils.PxSpecImages, t)
+
+		// Delete portworx oci-mon pods and validate they get re-deployed
```
Comment (piyush-nimbalkar) [inline test/integration_test/basic_test.go:296]: nit: Don't need to add the comment if we are already saying the same in the following comment.

File: test/integration_test/basic_test.go [hunk 12]
```diff
@@ -243,3 +270,79 @@ func testInstallWithTelemetry(t *testing.T, cluster *corev1.StorageCluster) {
...
+		logrus.Info("Delete portworx pods and validate they get re-deployed")
+		//err = testutil.DeletePodsByLabels(cluster.Namespace, map[string]string{"name": "portworx"})
```
Comment (piyush-nimbalkar) [inline test/integration_test/basic_test.go:298]: Remove this commented line

File: test/integration_test/basic_test.go [hunk 13]
```diff
@@ -243,3 +270,79 @@ func testInstallWithTelemetry(t *testing.T, cluster *corev1.StorageCluster) {
...
+		cluster, err = ci_utils.UpdateStorageCluster(cluster)
+		require.NoError(t, err)
+
+		// Sleep for 20 seconds to let operator start the update process
+		logrus.Debug("Sleeping for 20 seconds...")
+		time.Sleep(20 * time.Second)
+
+		// Validate cluster deployment
+		logrus.Infof("Validate StorageCluster %s", cluster.Name)
+		err = testutil.ValidateStorageCluster(ci_utils.PxSpecImages, cluster, ci_utils.DefaultValidateUpgradeTimeout, ci_utils.DefaultValidateDeployRetryInterval, true, "")
```
Comment (piyush-nimbalkar) [inline test/integration_test/basic_test.go:325]: I guess you wanted to use `ci_utils.DefaultValidateDeployTimeout` and not `ci_utils.DefaultValidateUpgradeTimeout`.

File: pkg/util/test/util.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (piyush-nimbalkar) [reply pkg/util/test/util.go:973]: This is still not using `GetDeploymentPods()`

File: test/integration_test/basic_test.go (hunk 13 above)
Comment (nikolaypopov) [reply test/integration_test/basic_test.go]: Done, made a new timeout vars for Update

File: test/integration_test/basic_test.go (hunk 12 above)
Comment (nikolaypopov) [reply test/integration_test/basic_test.go]: Removed

File: test/integration_test/basic_test.go (hunk 11 above)
Comment (nikolaypopov) [reply test/integration_test/basic_test.go]: Remove this and other obvious comments

File: test/integration_test/basic_test.go [hunk 14]
```diff
@@ -243,3 +270,79 @@ func testInstallWithTelemetry(t *testing.T, cluster *corev1.StorageCluster) {
 	err = coreops.Instance().DeleteSecret(secret.Name, secret.Namespace)
 	require.NoError(t, err)
 }
+
+// BasicCsiRegression test includes the following steps:
+// 1. Deploy PX with CSI enabled and validate CSI components and images
+// 2. Delete "px-csi-ext" pods and validate they get re-deployed
+// 3. Disable CSI and validate CSI components got successfully removed
+// 4. Enabled CSI and validate CSI components and images
+// 5. Delete "portworx" pods and validate they get re-deployed
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply test/integration_test/basic_test.go]: Update with correct steps

File: pkg/util/test/util.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Done

**Review by piyush-nimbalkar — APPROVED** [review]

**Pull request merged**